reference-implementation/lib/transform-stream.js

```diff
@@ -262,12 +238,32 @@ class TransformStreamSource {
 
     transformStream._readableController = c;
 
-    return this._startPromise;
+    // delay the first source.pull call until there is backpressure
+    return this._startPromise.then(() => {
+      if (transformStream._readableBackpressure === true) {
+        return true;
+      }
+
+      transformStream._backpressureChangePromise.then(backpressure => assert(backpressure === true));
+
+      return transformStream._backpressureChangePromise;
+    });
   }
 
   pull() {
-    this._transformStream._readableBackpressure = false;
-    TransformStreamTransformIfNeeded(this._transformStream);
+    // console.log('TransformStreamSource.pull()');
+
+    const transformStream = this._transformStream;
+    // invariant: pull is never called while readableBackpressure is false.
+    // Enforced by the promises returned by source.start and source.pull.
+    assert(transformStream._readableBackpressure === true);
+    assert(transformStream._backpressureChangePromise !== undefined);
+
+    TransformStreamSetBackpressure(transformStream, false);
+
+    transformStream._backpressureChangePromise.then(backpressure => assert(backpressure === true));
+
+    return transformStream._backpressureChangePromise;
```

Comment:
This is looping back the backpressure signal on desiredSize to pull() but mux-ed with backpressure-changing TransformStreamEnqueueToReadable() invocation. I think we shouldn't have to have such a loop back here. Let me think more.